src/triage/component/catwalk/evaluation.py

```diff
@@ -1,24 +1,77 @@
 import functools
+import io
 import logging
 import time
+import warnings
 
 import numpy
+import pandas
 from sqlalchemy.orm import sessionmaker
 
 from . import metrics
-from .utils import db_retry, sort_predictions_and_labels
+from .utils import (
+    db_retry,
+    sort_predictions_and_labels,
+    get_subset_table_name,
+    filename_friendly_hash
+)
+
+
+def subset_labels_and_predictions(
+    subset_df,
+    labels,
+    predictions_proba,
+):
+    indexed_predictions = pandas.Series(predictions_proba, index=labels.index)
+    
+    # The subset isn't specific to the cohort, so inner join to the labels/predictions
+    labels_subset = labels.align(subset_df, join="inner")[0]
+    predictions_subset = indexed_predictions.align(subset_df, join="inner")[0].values
+
+    logging.debug(
+        "%s entities in subset out of %s in matrix.",
+        len(labels_subset),
+        len(labels),
+    )
+
+    return (labels_subset, predictions_subset)
+
+
+def query_subset_table(db_engine, as_of_dates, subset_table_name):
+    as_of_dates = [date.strftime("%Y-%m-%d %H:%M:%S.%f") for date in as_of_dates],
+    query_string = f"""
+        with dates as (
+            select unnest(array{list(as_of_dates)}::timestamp[]) as as_of_date
+        )
+        select entity_id, as_of_date, active
+        from {subset_table_name}
+        join dates using(as_of_date)
+    """
+    copy_sql = f"COPY ({query_string}) TO STDOUT WITH CSV HEADER"
+    conn = db_engine.raw_connection()
+    cur = conn.cursor()
+    out = io.StringIO()
+    logging.debug(f"Running query {copy_sql} to get subset")
+    cur.copy_expert(copy_sql, out)
+    out.seek(0)
+    df = pandas.read_csv(
+        out, parse_dates=["as_of_date"],
+        index_col=["entity_id", "as_of_date"]
+    )
```

Comment:
…In fact, you could now:

```
$ pip install ohio
```

And then, here, this whole thing could be:

```
copy_sql = 'blah blah blah'
logging.debug("Running query %s to get subset", copy_sql)
conn = db_engine.raw_connection()
try:
    cur = conn.cursor()
    writer = lambda pipe: cur.copy_expert(copy_sql, pipe)
    with ohio.PipeTextIO(writer) as pipe:
        return pandas.read_csv(
            pipe,
            parse_dates=["as_of_date"],
            index_col=["entity_id", "as_of_date"]
        )
finally:
    conn.close()
```

(It would still be nicer with a context manager's `with` statement in place of the enclosing `try`/`except`; but, I haven't checked that the raw psql connection supports this.)

And, anyway, no need to manage a `StringIO`, and only as much is held in memory as required by the components of the pipe, (and we should still get all the benefits of a `copy`). …At least, that was the idea!